packages/libs/error-stack/tests/common.rs

```diff
@@ -175,3 +179,132 @@ pub fn messages<E>(report: &Report<E>) -> Vec<String> {
 pub fn frame_kinds<E>(report: &Report<E>) -> Vec<FrameKind> {
     report.frames().map(Frame::kind).collect()
 }
+
+#[cfg(all(nightly, feature = "std"))]
+pub fn supports_backtrace() -> bool {
+    // we need to track 3 states:
+    //  1) -1 (not checked)
+    //  2) 0 (disabled)
+    //  3) 1 (enabled)
+    static STATE: AtomicI8 = AtomicI8::new(-1);
+
+    match STATE.load(Ordering::SeqCst) {
+        -1 => {
+            let bt = std::backtrace::Backtrace::capture();
+            if bt.status() == std::backtrace::BacktraceStatus::Captured {
+                STATE.store(1, Ordering::SeqCst);
+                true
+            } else {
+                STATE.store(0, Ordering::SeqCst);
+                false
+            }
+        }
+        0 => false,
+        1 => true,
+        _ => unreachable!(),
+    }
+}
+
+#[cfg(feature = "spantrace")]
+pub fn supports_spantrace() -> bool {
+    static STATE: AtomicI8 = AtomicI8::new(-1);
+
+    match STATE.load(Ordering::SeqCst) {
+        -1 => {
+            let st = tracing_error::SpanTrace::capture();
+            if st.status() == tracing_error::SpanTraceStatus::CAPTURED {
+                STATE.store(1, Ordering::SeqCst);
+                true
+            } else {
+                STATE.store(0, Ordering::SeqCst);
+                false
+            }
+        }
+        0 => false,
+        1 => true,
+        _ => unreachable!(),
+    }
+}
+
+/// Conditionally add two opaque layers to the end,
+/// as these catch the backtrace and spantrace if recorded.
+pub fn expect_messages<'a>(messages: &[&'a str]) -> Vec<&'a str> {
+    #[allow(unused_mut)]
+    let mut messages = alloc::vec::Vec::from(messages);
+    // the last entry should always be `Context`, `Backtrace` and `Spantrace`
+    // are both added after the `Context` therefore we need to push those layers, then `Context`
+    let last = messages.pop().unwrap();
+
+    #[cfg(all(nightly, feature = "std"))]
+    if supports_backtrace() {
+        messages.push("Opaque");
+    }
+
+    #[cfg(feature = "spantrace")]
+    if supports_spantrace() {
+        messages.push("Opaque");
+    }
```

Comment:
Is this required because multiple traces are now a thing, so `zip`ping does not get rid of them?

Reply:
Yep, that's precisely why. This worked before (before one of my reworks), where we would attach the backtrace and spantrace before the context. But, because now those are added before the `Context,` we have a problem because now we have two conditional Opaque layers just before the root layer and can no longer easily use `zip`.

This is why `expect_messages`, `expect_count` and `assert_kinds!` exists.

Example:
The actual layer is:
`["Printable A", "Printable B", "Opaque"?, "Opaque"?, "Context"]` where the Opaque layer is only added if backtraces are enabled and can be captured. I didn't want to repeat the logic (shown above) every single time, so I create functions to accommodate for that issue.

Reply:
That makes sense, thanks for the explanation!